### PR TITLE
Fall back to hex selector if missing

### LINKED ISSUE #4058: Fix mocked selector in debugging and gas report
### Select Tool

snforge

### Foundry Version

0.55.0-rc.0

### What operating system are you using?

MacOS

### What system architecture are you using?

arm

### Issue Description

https://github.com/foundry-rs/starknet-foundry/blob/b79f597b6b154a24dc0edbeef5a707739c6151f4/crates/debugging/src/trace/collect.rs#L94
https://github.com/foundry-rs/starknet-foundry/blob/b79f597b6b154a24dc0edbeef5a707739c6151f4/crates/forge-runner/src/gas/report.rs#L156

It seems to fail when using mock calls with existing class hash, but non existent selector. To reproduce run `snforge test --gas-report` in `cairo-contracts/packages/governance` on https://github.com/OpenZeppelin/cairo-contracts/tree/1e5a44ec50d7625395ca8322eec971ce8abb5bc1

### Command Line Output

```bash
thread 'tokio-runtime-worker' (19740434) panicked at crates/forge-runner/src/gas/report.rs:156:10:
`Selector` should be present
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[ERROR] task 1112 panicked with message "`Selector` should be present"
```

### LINKED ISSUE #4065: Forge debugging fails in case of catching errors in nested calls
### Current State

Running a provided reproduction causes the expect to fail with `calldata should be successfully transformed: FunctionNotFound(0x32564d7e0fe091d49b4c20f4632191e4ed6986bf993849879abfef9465def25)`.

```patch
Index: crates/forge/tests/data/debugging/src/lib.cairo
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/forge/tests/data/debugging/src/lib.cairo b/crates/forge/tests/data/debugging/src/lib.cairo
--- a/crates/forge/tests/data/debugging/src/lib.cairo	(revision d1aab69c75e571479b4e5a52003c2afe2a9699fd)
+++ b/crates/forge/tests/data/debugging/src/lib.cairo	(revision aabbca7ce94b240f491919f28fdbcca824a73d90)
@@ -16,11 +16,33 @@
     fn fail(self: @TContractState, data: Array<felt252>);
 }
 
+#[starknet::interface]
+trait Nested<TContractState> {
+    fn nested(ref self: TContractState, contract_address: ContractAddress);
+}
+
+
+#[starknet::contract]
+mod FailingContract {
+    use crate::Failing;
+
+    #[storage]
+    struct Storage {}
+
+
+    impl FailingImpl of Failing<ContractState> {
+        fn fail(self: @ContractState, data: Array<felt252>) {
+            panic_with_felt252('panicked');
+        }
+    }
+}
+
 #[starknet::contract]
 mod SimpleContract {
     use core::array::ArrayTrait;
     use core::traits::Into;
     use starknet::{ContractAddress, get_contract_address};
+    use crate::{FailingSafeDispatcher, FailingSafeDispatcherTrait, Nested};
     use super::{
         Failing, RecursiveCall, RecursiveCaller, RecursiveCallerDispatcher,
         RecursiveCallerDispatcherTrait,
@@ -57,4 +79,15 @@
             panic(data);
         }
     }
+
+    #[abi(embed_v0)]
+    impl NestedImpl of Nested<ContractState> {
+        #[feature("safe_dispatcher")]
+        fn nested(ref self: ContractState, contract_address: ContractAddress) {
+            let dispatcher = FailingSafeDispatcher { contract_address };
+            if let Ok(_) = dispatcher.fail(array![]) {
+                panic!("should have panicked")
+            }
+        }
+    }
 }
Index: crates/forge/tests/data/debugging/tests/test_trace.cairo
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/crates/forge/tests/data/debugging/tests/test_trace.cairo b/crates/forge/tests/data/debugging/tests/test_trace.cairo
--- a/crates/forge/tests/data/debugging/tests/test_trace.cairo	(revision d1aab69c75e571479b4e5a52003c2afe2a9699fd)
+++ b/crates/forge/tests/data/debugging/tests/test_trace.cairo	(revision aabbca7ce94b240f491919f28fdbcca824a73d90)
@@ -1,6 +1,6 @@
 use debugging::{
-    FailingDispatcher, FailingDispatcherTrait, RecursiveCall, RecursiveCallerDispatcher,
-    RecursiveCallerDispatcherTrait,
+    FailingDispatcher, FailingDispatcherTrait, NestedDispatcher, NestedDispatcherTrait,
+    RecursiveCall, RecursiveCallerDispatcher, RecursiveCallerDispatcherTrait,
 };
 use snforge_std::cheatcodes::contract_class::DeclareResultTrait;
 use snforge_std::trace::get_call_trace;
@@ -37,6 +37,14 @@
 
     RecursiveCallerDispatcher { contract_address: contract_address_A }.execute_calls(calls);
 
+    let nested_dispatcher = NestedDispatcher { contract_address: contract_address_A };
+    let (failing_contract, _) = declare("FailingContract")
+        .unwrap()
+        .contract_class()
+        .deploy(@array![])
+        .unwrap();
+    nested_dispatcher.nested(failing_contract);
+
     let failing_dispatcher = FailingDispatcher { contract_address: contract_address_A };
     failing_dispatcher.fail(array![1, 2, 3, 4, 5]);
 }
```

### Objective

Fix the issue

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ddoktorski**
Fixes #4058
Fixes #4065

- Fall back to raw selector in forge debugging and gas report in case it is missing in ABI

### CODE REVIEW

**Comment by ddoktorski** [comment]
This change is part of the following stack:

- #4325 ◀
    - #4326

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: crates/forge/tests/e2e/debugging.rs [hunk 1]
```diff
@@ -116,6 +116,37 @@ fn debugging_trace_minimal_fork() {
...
+
+            Collected 1 test(s) from debugging_no_abi package
+            Running 0 test(s) from src/
+            Running 1 test(s) from tests/
+            [PASS] debugging_no_abi_integrationtest::test_trace::test_nested_safe_call_no_abi (l1_gas: ~0, l1_data_gas: ~[..], l2_gas: ~[..])
+            [test name] debugging_no_abi_integrationtest::test_trace::test_nested_safe_call_no_abi
+            └─ [selector] 0x32564d7e0fe091d49b4c20f4632191e4ed6986bf993849879abfef9465def25
+               ├─ [contract name] CallerContract
+               ├─ [calldata] 0x0
+               └─ [call result] success
```
Comment (ddoktorski) [inline crates/forge/tests/e2e/debugging.rs:143]: Note: This trace is incorrect for now. It is related to a different issue, which is addressed in next PR (#4326)

File: crates/debugging/src/trace/collect.rs [hunk 2]
```diff
@@ -101,26 +107,31 @@ impl<'a> Collector<'a> {
...
-        )
+        let calldata = &self.call_trace.entry_point.calldata.0;
+        let selector = &self.call_trace.entry_point.entry_point_selector.0;
+        let transformed = reverse_transform_input(calldata, abi, selector).unwrap_or_else(|_| {
+            calldata
+                .iter()
+                .map(|f| format!("{f:#x}"))
+                .collect::<Vec<_>>()
+                .join(", ")
+        });
```
Comment (franciszekjob) [inline crates/debugging/src/trace/collect.rs]: This fallback looks a bit too broad because it now catches any `ReverseTransformError`, not just the missing-selector case (so for example any decode mismatches will not be surfaced). Is this intended?

File: crates/debugging/src/trace/collect.rs [hunk 3]
```diff
@@ -101,26 +107,31 @@ impl<'a> Collector<'a> {
...
-                .expect("call result should be successfully transformed");
-                format_result_message("success", &ret_data)
+                let ret_data_str = reverse_transform_output(ret_data, abi, selector)
+                    .unwrap_or_else(|_| {
+                        ret_data
+                            .iter()
+                            .map(|f| format!("{f:#x}"))
+                            .collect::<Vec<_>>()
+                            .join(", ")
+                    });
```
Comment (franciszekjob) [inline crates/debugging/src/trace/collect.rs]: nit: This code is the same as above (in `collect_transformed_calldata`), so we can add a small helper:

```suggestion
                    .unwrap_or_else(|_| format_raw_felts(calldata));
```

```rust
fn format_raw_felts(felts: &[Felt]) -> String {
    felts
        .iter()
        .map(|felt| format!("{felt:#x}"))
        .collect::<Vec<_>>()
        .join(", ")
}
```

File: crates/debugging/src/trace/collect.rs
Comment (franciszekjob) [inline crates/debugging/src/trace/collect.rs:1]: nit: Could we add tests with non-empty calldata and return data here as well? That would verify the actual raw-hex fallback formatting.

File: crates/debugging/src/trace/collect.rs (hunk 2 above)
Comment (ddoktorski) [reply crates/debugging/src/trace/collect.rs]: Agree it's cleaner to fall back only when there is a relevant error, changed

**Review by franciszekjob — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Forge
+
+#### Fixed
+
+- Bug in forge debugging and `--gas-report` that caused panic in case of a call to non-existent selector or malformed calldata
```
Comment (DelevoXDG) [inline CHANGELOG.md]: Is the "calldata" part still true following 2e9c8b0b9?
```suggestion
- Bug in forge debugging and `--gas-report` that caused panic in case of a call to non-existent selector
```

**Review by DelevoXDG — APPROVED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (ddoktorski) [reply CHANGELOG.md]: Done
